File: core/entity/warrior.js
"use strict";

var obj = function(map, config) {
    this.map = map;
    this.type = 'warrior';
    this.pos = {
        x: config.x || 0,
        y: config.y || 0
    };
    this.health = 100;
    this.pulse = 0;
    this.nexus = config.nexus;
    this.side = config.side;
};
obj.prototype = require(appRoot + '/core/lib/extends.js')(require(appRoot + '/core/entity.js'), {
    think: function() {
        this.createPulse('nexus_' + this.side);

        var cord = [[-1, 0], [1, 0], [0, -1], [0, 1]];

        var score = 0, next = null;
        for (var v in cord) {
            var c = cord[v];

            for (var i in this.map.entity) {
                var e = this.map.entity[i];
                if (e.side != this.side && (e.type == 'worker' || e.type == 'warrior' || e.type == 'nexus') && e.pos.x == this.pos.x + c[0] && e.pos.y == this.pos.y + c[1]) {
                    if (this.map.entity[i].health > 0) {
                        this.map.entity[i].health -= Math.floor(Math.random() * 2);
                    } else {
                        this.map.entity.splice(i, 1);
                    }
                    return;
                }
            }

            for (var k in this.map.side) {
                if (this.map.side[i] != this.side) {
                    var f = this.heat('nexus_' + this.map.side[k], this.pos.x + c[0], this.pos.y + c[1]);
                    if (f > score) {
                        score = f;
                        next = [this.pos.x + c[0], this.pos.y + c[1]];
                    }
                }
            }
        }

        if (next) {
            this.setPos(next[0], next[1]);
            this.map.heat['nexus_' + this.side].set(this.pos.x, this.pos.y, 0);
        } else {
            var c = cord[Math.floor(Math.random() * cord.length)];
            this.setPos(this.pos.x + c[0], this.pos.y + c[1]);
        }
    }
});

module.exports = obj;

File: java/Warrior.java

public class Warrior extends Entity implements IEntity {
	private Map map;
	public Nexus nexus;

	protected Warrior(Map m, Nexus n, float x, float y) {
		this.map = m;
		this.nexus = n;
		this.x = x;
		this.y = y;
		log.write("log/game", "spawn warrior " + n.side);
	}

	private Boolean attack(Entity b) {
		if (b.x == Math.round(x) && b.y == Math.round(y)) {
			if (Seed.random(2) == 1) {
				return (true);
			}
		} else {
			nextPos(1, b.x, b.y);
		}
		return (false);
	}
	
	public Boolean think() {
		for(Object i : map.entity) {
			if (i instanceof Worker) {
				Worker a = (Worker) i;
				if (a.nexus != nexus) {
					if (attack(a)) {
						map.entity.remove(i);
						return (false);
					}
					return (true);
				}
			}
		}

		for(Object i : map.entity) {
			if (i instanceof Warrior) {
				Warrior b = (Warrior) i;
				if (b.nexus != nexus) {
					if (attack(b)) {
						map.entity.remove(i);
						return (false);
					}
					return (true);
				}
			}
		}

		for(Object i : map.entity) {
			if (i instanceof Nexus) {
				Nexus c = (Nexus)i;
				if (c != nexus) {
					if (c.x == Math.round(x) && c.y == Math.round(y)) {
						if (c.health > 0) {
							c.health -= 1;
						} else {
							map.entity.remove(i);
							return (false);
						}
					} else {
						nextPos(1, c.x, c.y);
					}
					return (true);
				}
			}
		}

		nextPos(1, x + (Seed.random(20) - 10), y + (Seed.random(20) - 10));
		return (true);
	}
}
